File: problems/LeetCode/p152.py
class Solution:
    def maxProduct(self, nums: int) -> int:
        res = max(nums)
        cu_min, cu_max = 1, 1

        for n in nums:
            if n == 0:
                cu_min, cu_max = 1, 1
                continue
            
            temp = n*cu_max
            cu_max = max(n*cu_max, n*cu_min, n) #some language doesn's allow [-1, 8]
            cu_min = min(temp, n*cu_min, n) # [-1, -8], we want -8
            res = max(res, cu_max)

        return res

File: LeetCode-Java/p152.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int curMin = 1, curMax = 1;
        
        for (int n : nums) {
            if (n == 0) {
                curMin = 1;
                curMax = 1;
                continue;
            }
            
            int temp = n * curMax;
            curMax = Math.max(n * curMax, n * curMin);
            curMax = Math.max(curMax, n);
            curMin = Math.min(temp, n * curMin);
            curMin = Math.min(curMin, n);
            res = Math.max(res, curMax);
        }
        
        return res;
    }
}
